File: OneCodeTeam/How-to-achieve-breakpoint-resume-when-upload-big-111754/[VB]-How to achieve breakpoint resume when upload big file to blob (VS2013)/VB/VBAzureUploadBigFile/Form1.vb
'**************************** Module Header ******************************\
' Module Name:	Form1.vb
' Project:		VBAzureUploadBigFile
' Copyright (c) Microsoft Corporation.
' 
' This sample demonstrates how to upload a big file to azure blob storage.
'
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*************************************************************************
Imports Microsoft.WindowsAzure.Storage.Blob
Imports Microsoft.WindowsAzure.Storage
Imports System.ComponentModel
Imports System.IO
Imports System.Text

Partial Public Class Form1
    Inherits Form
    Private storageConnectionString As String = "DefaultEndpointsProtocol=https;AccountName={Account-Name};AccountKey={Account-Key}"

    Private bufferSize As Integer = 40 * 1024
    Private blockCount As Integer = 0
    Private blockIds As New List(Of String)()
    Private filePath As String
    Private isCanceled As Boolean = False
    Public Sub New()
        InitializeComponent()
        btnStart.Enabled = False
        btnEnd.Enabled = False
    End Sub

    Private Sub uploadBackgroundWorker_DoWork(sender As Object, e As DoWorkEventArgs)
        UploadBigFile(uploadBackgroundWorker, e)
    End Sub

    Private Sub uploadBackgroundWorker_ProgressChanged(sender As Object, e As ProgressChangedEventArgs)
        progressBar1.Value = e.ProgressPercentage
    End Sub

    Private Sub uploadBackgroundWorker_RunWorkerCompleted(sender As Object, e As RunWorkerCompletedEventArgs)
        If Not isCanceled Then
            blockIds.Clear()
        End If
        btnStart.Enabled = True
        browseFile.Enabled = True
        btnEnd.Enabled = False
    End Sub

    Private Sub browseFile_Click(sender As Object, e As EventArgs)
        If openFileDialog1.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
            textBox1.Text = InlineAssignHelper(filePath, openFileDialog1.FileName)
            Using fileStream As FileStream = File.OpenRead(filePath)
                blockCount = CInt(fileStream.Length / bufferSize) + 1
            End Using
            progressBar1.Maximum = blockCount
        End If
        If Not String.IsNullOrEmpty(textBox1.Text) Then
            btnStart.Enabled = True
        End If
    End Sub

    Private Sub btnStart_Click(sender As Object, e As EventArgs)
        If Not uploadBackgroundWorker.IsBusy Then
            btnStart.Enabled = False
            browseFile.Enabled = False
            btnEnd.Enabled = True
            uploadBackgroundWorker.RunWorkerAsync()
        End If
    End Sub

    Private Sub btnEnd_Click(sender As Object, e As EventArgs)
        btnStart.Enabled = True
        browseFile.Enabled = True
        btnEnd.Enabled = False
        uploadBackgroundWorker.CancelAsync()
        isCanceled = True
    End Sub

    Private Sub UploadBigFile(worker As BackgroundWorker, e As DoWorkEventArgs)
        'bufferSize 40KB
        Dim bufferBytes As Byte() = New Byte(bufferSize - 1) {}
        Dim fileName As String = Path.GetFileName(filePath)

        Dim blob As CloudBlockBlob = GetBlobkBlob(fileName)

        Using fileStream As FileStream = File.OpenRead(filePath)
            'Get the total count of block
            blockCount = CInt(fileStream.Length / bufferSize) + 1
            Dim currentBlockSize As Int64 = 0
            Dim currentCount As Integer = blockIds.Count()
            fileStream.Seek(bufferSize * currentCount, SeekOrigin.Begin)
            For i As Integer = blockIds.Count To blockCount - 1
                If worker.WorkerSupportsCancellation AndAlso worker.CancellationPending Then
                    Return
                End If
                currentBlockSize = bufferSize
                If i = blockCount - 1 Then
                    currentBlockSize = fileStream.Length - bufferSize * i
                    bufferBytes = New Byte(currentBlockSize - 1) {}
                End If
                If currentBlockSize = 0 Then
                    Exit For
                End If
                'Get the block data
                fileStream.Read(bufferBytes, 0, Convert.ToInt32(currentBlockSize))
                Using memoryStream As New MemoryStream(bufferBytes)
                    Try
                        'Get Base64-encoded block Id
                        Dim blockId As String = Convert.ToBase64String(Encoding.UTF8.GetBytes(Guid.NewGuid().ToString()))
                        'Upload the block
                        blob.PutBlock(blockId, memoryStream, Nothing)
                        'Cache the block Id
                        blockIds.Add(blockId)
                        worker.ReportProgress(i + 1)

                    Catch generatedExceptionName As Exception
                    End Try
                End Using
            Next
        End Using
        'Commit all the blocks
        blob.PutBlockList(blockIds)
        isCanceled = False
    End Sub

    Private Function GetBlobkBlob(fileName As String) As CloudBlockBlob
        Dim storageAccount = CloudStorageAccount.Parse(storageConnectionString)
        Dim blobStorage As CloudBlobClient = storageAccount.CreateCloudBlobClient()

        Dim container As CloudBlobContainer = blobStorage.GetContainerReference("mycontainer")
        container.CreateIfNotExists()

        Return container.GetBlockBlobReference(fileName)
    End Function
    Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
        target = value
        Return value
    End Function
End Class

File: OneCodeTeam/How-to-achieve-breakpoint-resume-when-upload-big-111753/[C#]-How to achieve breakpoint resume when upload big file to blob (VS2012)/CSAzureUploadBigFile/CSAzureUploadBigFile/Form1.cs
/***************************** Module Header ******************************\
 Module Name:	Form1.cs
 Project:		CSAzureUploadBigFile
 Copyright (c) Microsoft Corporation.
 
 This sample demonstrates how to upload a big file to azure blob storage.

 This source is subject to the Microsoft Public License.
 See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
 All other rights reserved.
 
 THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
 EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
 WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
**************************************************************************/
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Blob;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace CSAzureUploadBigFile
{
    public partial class Form1 : Form
    {
        private string storageConnectionString
    = "DefaultEndpointsProtocol=https;AccountName={account-name};AccountKey={account-key}";

        private int bufferSize = 40 * 1024;
        private int blockCount = 0;
        private List<string> blockIds = new List<string>();
        private string filePath;
        private bool isCanceled = false;
        public Form1()
        {
            InitializeComponent();
            btnStart.Enabled = false;
            btnEnd.Enabled = false;
        }

        private void uploadBackgroundWorker_DoWork(object sender, DoWorkEventArgs e)
        {
            UploadBigFile(uploadBackgroundWorker, e);
        }

        private void uploadBackgroundWorker_ProgressChanged(object sender, ProgressChangedEventArgs e)
        {
            progressBar1.Value = e.ProgressPercentage;
        }

        private void uploadBackgroundWorker_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            if (!isCanceled)
            {
                blockIds.Clear();
            }
            btnStart.Enabled = true;
            browseFile.Enabled = true;
            btnEnd.Enabled = false;
        }

        private void browseFile_Click(object sender, EventArgs e)
        {
            if (openFileDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK)
            {
                textBox1.Text = filePath = openFileDialog1.FileName;
                using (FileStream fileStream = File.OpenRead(filePath))
                {
                    blockCount = (int)(fileStream.Length / bufferSize) + 1;
                }
                progressBar1.Maximum = blockCount;
            }
            if (!string.IsNullOrEmpty(textBox1.Text))
            {
                btnStart.Enabled = true;
            }
        }

        private void btnStart_Click(object sender, EventArgs e)
        {
            if (!uploadBackgroundWorker.IsBusy)
            {
                btnStart.Enabled = false;
                browseFile.Enabled = false;
                btnEnd.Enabled = true;
                uploadBackgroundWorker.RunWorkerAsync();
            }
        }

        private void btnEnd_Click(object sender, EventArgs e)
        {
            btnStart.Enabled = true;
            browseFile.Enabled = true;
            btnEnd.Enabled = false;
            uploadBackgroundWorker.CancelAsync();
            isCanceled = true;
        }

        private void UploadBigFile(BackgroundWorker worker, DoWorkEventArgs e)
        {
            //bufferSize 40KB
            byte[] bufferBytes = new byte[bufferSize];
            string fileName = Path.GetFileName(filePath);

            CloudBlockBlob blob = GetBlobkBlob(fileName);

            using (FileStream fileStream = File.OpenRead(filePath))
            {
                //Get the total count of block
                blockCount = (int)(fileStream.Length / bufferSize) + 1;
                Int64 currentBlockSize = 0;
                int currentCount = blockIds.Count();
                fileStream.Seek(bufferSize * currentCount, SeekOrigin.Begin);
                for (int i = blockIds.Count; i < blockCount; i++)
                {
                    if (worker.WorkerSupportsCancellation && worker.CancellationPending)
                    {
                        return;
                    }
                    currentBlockSize = bufferSize;
                    if (i == blockCount - 1)
                    {
                        currentBlockSize = fileStream.Length - bufferSize * i;
                        bufferBytes = new byte[currentBlockSize];
                    }
                    if (currentBlockSize == 0) break;
                    //Get the block data
                    fileStream.Read(bufferBytes, 0, Convert.ToInt32(currentBlockSize));
                    using (MemoryStream memoryStream = new MemoryStream(bufferBytes))
                    {
                        try
                        {
                            //Get Base64-encoded block Id
                            string blockId = Convert.ToBase64String(Encoding.UTF8.GetBytes(Guid.NewGuid().ToString()));
                            //Upload the block
                            blob.PutBlock(blockId, memoryStream, null);
                            //Cache the block Id
                            blockIds.Add(blockId);
                            worker.ReportProgress(i + 1);
                        }
                        catch (Exception)
                        {

                        }
                    }
                }
            }
            //Commit all the blocks
            blob.PutBlockList(blockIds);
            isCanceled = false;
        }

        private CloudBlockBlob GetBlobkBlob(string fileName)
        {
            var storageAccount = CloudStorageAccount.Parse(storageConnectionString);
            CloudBlobClient blobStorage = storageAccount.CreateCloudBlobClient();

            CloudBlobContainer container = blobStorage.GetContainerReference("mycontainer");
            container.CreateIfNotExists();

            return container.GetBlockBlobReference(fileName);
        }
    }
}
